be/src/formats/parquet/meta_helper.cpp

```diff
@@ -53,16 +42,64 @@ void ParquetMetaHelper::build_column_name_2_pos_in_meta(
 }
 
 void ParquetMetaHelper::prepare_read_columns(const std::vector<HdfsScannerContext::ColumnInfo>& materialized_columns,
-                                             std::vector<GroupReaderParam::Column>& read_cols) const {
+                                             std::vector<GroupReaderParam::Column>& read_cols,
+                                             std::unordered_set<std::string>& existed_column_names) const {
     for (auto& materialized_column : materialized_columns) {
         int32_t field_idx = _file_metadata->schema().get_field_idx_by_column_name(materialized_column.name());
         if (field_idx < 0) continue;
 
-        auto parquet_type = _file_metadata->schema().get_stored_column_by_field_idx(field_idx)->physical_type;
+        const ParquetField* parquet_field = _file_metadata->schema().get_stored_column_by_field_idx(field_idx);
+        // check is complex type is invalid
+        if (parquet_field->type.is_complex_type() &&
+            !_is_valid_complex_type(parquet_field, &materialized_column.slot_desc->type())) {
+            continue;
+        }
+
+        auto parquet_type = parquet_field->physical_type;
         GroupReaderParam::Column column = _build_column(field_idx, parquet_type, materialized_column.slot_desc,
                                                         materialized_column.decode_needed);
         read_cols.emplace_back(column);
+        existed_column_names.emplace(materialized_column.name());
+    }
+}
+
+bool ParquetMetaHelper::_is_valid_complex_type(const ParquetField* parquet_field,
+                                               const TypeDescriptor* type_descriptor) const {
+    DCHECK(parquet_field->type.is_complex_type());
+    // now we will only check for struct type
+    // if struct has none valid subfield, we will treat this struct type as invalid type.
+    if (parquet_field->type.is_array_type() || parquet_field->type.is_map_type()) {
```

Comment:
i think we also need check child of array and map, because of we have prune on unnest, so there are two scenario, 1. there is no any leaf node, so the root node is empty, we mark it as not exist. 2. there is some medium node has not any left node, we must check and mark the medium column reader as nullptr, and because child reader may be nullptr, we must check it array, map, struct and mark parent as nullptr, and it must be subfield of struct or map, and can be filled by referencing it's brother node.

Reply:
fixed